extensions/amp-story/0.1/audio.js

```diff
@@ -0,0 +1,429 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+
+/**
+ * The volume to which audio should be reduced when other audio sources are
+ * played.
+ * @const {number}
+ */
+const REDUCED_VOLUME = 0.5;
+
+/**
+ * @const {number}
+ */
+const VOLUME_CHANGE_DURATION_MS = 500;
+
+/**
+ * @const {!function(number): number}
+ */
+const VOLUME_EASING_FN = input => input;
+
+/**
+ * @const {string}
+ */
+const PLAYABLE_ID_PREFIX = 'i-amphtml-playable-';
+
+
+/**
+ * @const {string}
+ */
+const BACKGROUND_AUDIO_ELEMENT_CLASS_NAME = 'i-amphtml-story-background-audio';
+
+
+/**
+ * Adds support for the background-audio property on the specified element.
+ * @param {!Element} element The element to upgrade with support for background
+ *     audio.
+ */
+export function upgradeBackgroundAudio(element) {
+  if (element.hasAttribute('background-audio')) {
+    const audioEl = element.ownerDocument.createElement('audio');
+    const audioSrc = element.getAttribute('background-audio');
+    audioEl.setAttribute('src', audioSrc);
+    audioEl.setAttribute('preload', 'auto');
+    audioEl.setAttribute('loop', '');
+    audioEl.setAttribute('autoplay', '');
+    audioEl.classList.add(BACKGROUND_AUDIO_ELEMENT_CLASS_NAME);
+    element.appendChild(audioEl);
+  }
+}
+
+
+export class AudioManager {
+  constructor(win, rootElement) {
+    /** @private @const {!Object<!Element, !Playable>} */
+    this.playables_ = {};
+
+    /** @private @const {!Array<!Playable>} */
+    this.nowPlaying_ = [];
+
+    /** @private {number} */
+    this.nextId_ = 0;
+
+    /** @private {boolean} */
+    this.isMuted_ = true;
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.rootElement_ = rootElement;
+  }
+
+  /**
+   * @param {!Element} sourceElement The element causing audio to be played.
+   * @return {!Playable} The {@link Playable} instance to play the audio
+   *     represented by the specified sourceElement.
+   */
+  createPlayable_(sourceElement) {
+    if (!(sourceElement instanceof Element)) {
+      dev().error('AMP-STORY', 'Played item must be element.');
+    }
+
+    if (sourceElement instanceof HTMLMediaElement) {
+      return new MediaElementPlayable(this.win_, sourceElement);
+    }
+  }
+
+  /**
+   * @param {!Element} sourceElement
+   * @return {?Playable}
+   */
+  getPlayable_(sourceElement) {
+    return this.playables_[sourceElement.id];
+  }
+
+  /**
+   * Loads the audio for the specified.
+   * @param {!Element} sourceElement The element whose audio should be loaded.
+   */
+  load(sourceElement) {
+    const playable = this.getPlayable_(sourceElement) ||
+        this.createPlayable_(sourceElement);
+
+    if (!playable) {
```

Comment:
If  `createPlayable_` ends up actually returning `!Playable` based on the comment above, this can go away.

Reply:
Done